Code restructuring part 1 : no globals

ihavnoid opened the pull request:
Restructured code to remove:
- No global / thread_local OpenCL structures
- Network is now all non-static methods with instance variables
- NNCache is a member of Network, OpenCLScheduler is also a member of Network
- ThreadData is renamed to OpenCLContext (it's no longer thread-specific!)
- Other accompanying fixes

ihavnoid commented:
I can't figure out why the test is failing - any help?

gcp commented:
Thanks! Give me a few moments to review this.

gcp commented on src/OpenCLScheduler.h:
```diff
@@ -36,22 +37,13 @@ class OpenCLScheduler {
-    Utils::ThreadPool m_threadpool;
+    // XXX : I failed to figure out how to make this compile with unique_ptr
+    // especially on visual studio
+    std::vector<std::list<std::shared_ptr<OpenCLContext>>> m_context;
```
I might make sense to make a helper type for the innermost list. Would also make it clearer why 2 types of collections are being nested here.

gcp commented on src/UCTSearch.cpp:
```diff
@@ -71,11 +71,23 @@ class OutputAnalysisData {
 
 
-UCTSearch::UCTSearch(GameState& g)
+UCTSearch::UCTSearch(GameState& g, const std::string & weightsfile)
```
Meh. I don't think UCTSearch should know about a string indicating a file for the network. Wouldn't it make more sense if it took a Network& instead? GTP:: should then inject that ref into search.

This avoids some ugliness like GameState::place_free_handicap taking an UCTSearch*, even though it doesn't want an existing UCTSearch at all, it just really needs the info to set up the network for it.

You can probably also avoid that "global" search object you have in GTP now. The network that is specified on the commandline can be loaded an injected from the main function and cmdline parsing then.

ihavnoid replied:
Yeah I agree.  Trying to remove the global Network reference is causing more trouble than necessary, making other places looking odd for no apparent benefit.  I'll move the network variable to GTP and just pass a reference.

ihavnoid commented on src/GTP.cpp:
```diff
@@ -76,6 +76,12 @@ std::string cfg_options_str;
 bool cfg_benchmark;
 int cfg_analyze_interval_centis;
 
+std::unique_ptr<UCTSearch> GTP::search;
```
You probably figured out, but this was due to an attempt trying to remove the global Network and keep it as a member of UCTSearch.  Without this, OpenCL gets initialized on first GTP command hence introducing some wierdness.  (Comment will continue below)

ihavnoid commented on src/OpenCLScheduler.cpp:
```diff
@@ -22,61 +22,72 @@
-    auto f = m_threadpool.add_task([this, &input, &output_pol, &output_val]{
-        m_networks[current_thread_gpu_num]->forward(input, output_pol, output_val);
-    });
+    m_networks[ctx->m_gpu_num]->forward(input, output_pol, output_val, *ctx);
```
Yes it is guaranteed, otherwise the condition var should not exit.  Will add assertion here.

gcp replied:
Ok I guess it's ensured by the predicate inside the condvar wait.

ihavnoid commented:
Addressed 1st round of comments.
- Network is no longer child of UCTSearch, it's a reference to GTP::network
- Added assert on ctx
- Changed 'new OpenCLContext' to std::make_shared
- Added subtype for vector-of-list-of-pointer-of-OpenCLContext for readibility

gcp commented on src/OpenCLScheduler.h:
```diff
@@ -36,22 +37,16 @@ class OpenCLScheduler {
-};
 
-extern OpenCLScheduler opencl;
+    // XXX : I failed to figure out how to make this compile with unique_ptr
```
Comment can go, can always try to fix it later.

gcp commented on src/OpenCLScheduler.cpp:
```diff
@@ -22,61 +22,75 @@
-            }
-        }
-    } else {
+    for (auto gpu : gpus) {
```
const auto

gcp commented on src/OpenCLScheduler.cpp:
```diff
@@ -22,61 +22,75 @@
+    {
+        std::unique_lock<std::mutex> lk(m_context_lock);
+        m_context_condvar.wait(lk, [this]{
+            for (auto & ctxlist : m_context) {
```
const auto&

gcp commented on src/OpenCLScheduler.cpp:
```diff
@@ -22,61 +22,75 @@
+            }
+            return false;
+        });
+        while (queue_num < m_context.size()) {
```
Can't this be a for loop indexing over m_context? Ok, I see that you need queue_num later.

gcp commented on src/OpenCLScheduler.h:
```diff
@@ -36,22 +37,16 @@ class OpenCLScheduler {
+    // XXX : I failed to figure out how to make this compile with unique_ptr
+    // especially on visual studio
+    using OpenCLContextQueue = std::list<std::shared_ptr<OpenCLContext>>;
+    std::vector<OpenCLContextQueue> m_context;
```
Really a list of contexts per worker thread. m_work_threads? m_thread_contexts? Something else? Not sure.

ihavnoid replied:
Well, it's the 'pool' for OpenCL contexts.  Hence I just put it m_context_pool.  Not perfect but I don't think we should call this 'threads' anymore to avoid confusion (although yes it still has a lot of sense of the concept 'thread' here)

gcp commented on src/OpenCLScheduler.cpp:
```diff
@@ -22,61 +22,75 @@
-
+        opencl->initialize(channels, {gpu}, silent);
         m_opencl.push_back(std::move(opencl));
         m_networks.push_back(std::move(net));
```
Does this work correctly with --gpu 1 --gpu 3?

m_networks will be a size 2 array, yet ctx->m_gpu_num which is indexing into that can be 1 and 3.

ihavnoid replied:
Ouch x100.  Will rewrite code.

ihavnoid replied:
While trying to rewrite code, figured out that this isn't a bug - it is correct because m_gpu_num will be 0 or 1 (that is, this isn't the GPU number given by the command line argument.  This is the GPU index number on OpenCLScheduler.  Looking back, I don't like defining variables that are only used on OpenCLScheduler on a OpenCL data structure.  Will do a small rewrite to make it prettier...

gcp commented on src/Network.h:
```diff
@@ -79,43 +72,78 @@ class Network {
-};
 
+#ifdef USE_OPENCL
+    OpenCLScheduler m_opencl;
```
It feels a bit weird to have this owned by Network. I guess it's really here because it contains pointers to contexts which themselves identify where on the GPUs the network data is loaded.

ihavnoid replied:
I agree with you, but I couldn't find a better place for this.  I have another batch of fixes (abstract interface for network evaluation backend) and this is replaced with a pointer to the abstract type.  This piece needs a bit more clever thought but no good idea yet...

gcp commented on src/OpenCL.cpp:
```diff
@@ -412,26 +412,24 @@ const std::string sourceCode_sgemm =
-
-void OpenCL::ensure_thread_initialized() {
-    if (!opencl_thread_data.m_is_initialized) {
+void OpenCL::ensure_thread_initialized(OpenCLContext &opencl_context) {
```
If this takes a context argument it is no longer thread specific, so the function name is wrong.

ihavnoid replied:
renaming to ensure_thread_initialized

gcp commented on src/OpenCL.cpp:
```diff
@@ -1093,11 +1097,14 @@ void OpenCL::initialize(const int channels, const std::vector<int> & gpus,
 
-    ensure_thread_initialized();
+    // dummy gpu number, putting bogus number so that we don't misuse it
+    OpenCLContext tdata(0xffff);
```
I don't think you can misuse anything here given that OpenCLContext is no longer a thread local. Just make it 0 or whatever is the first in cfg_gpu. There's also nothing thread related any more here.

ihavnoid replied:
I will still leave the comment that the thread number here is bogus, just to prevent anybody from being confused...

ihavnoid commented:
No objection on any of your 2nd batch of comments.  Give me a couple of days to clean them up.

ihavnoid commented on src/GameState.cpp:
```diff
@@ -32,6 +32,7 @@
 #include "FullBoard.h"
 #include "KoState.h"
 #include "UCTSearch.h"
+#include "GTP.h"
```
Correct.  Will be gone.

ihavnoid commented on src/Network.h:
```diff
@@ -28,8 +28,13 @@
+#include "OpenCLScheduler.h"
+#endif
+
+class GameState;
```
...and due to this, we need to remove the \#include "Network.h" from GameState.h and replace it with a forward declaration.  Well, the change you suggested seems to be more logical..

ihavnoid commented:
Pushed 2nd batch of fixes.  Please have a look.

TFiFiE commented:
The following member functions of `Network` can stay **static**:
- `process_bn_var`
- `winograd_transform_f`
- `zeropad_U`
- `winograd_transform_in`
- `winograd_transform_out`
- `winograd_convolve3`
- `winograd_sgemm`

Also, `forward_cpu` can be declared **const** and the indentations of the function arguments in `Network.h` are in need of fixing.

ihavnoid commented:
Changed the functions' prototype to static, and fixed indentation.

TFiFiE commented on src/Network.h:
```diff
@@ -39,21 +43,9 @@ class Network {
+    Netresult get_scored_moves(const GameState* const state,
                                       const Ensemble ensemble,
                                       const int symmetry = -1,
                                       const bool skip_cache = false);
```
Indentation.

TFiFiE commented on src/Network.h:
```diff
@@ -67,8 +59,8 @@ class Network {
-    static void benchmark(const GameState * const state,
+    void initialize(int playouts, const std::string & weightsfile);
+    void benchmark(const GameState * const state,
                           const int iterations = 1600);
```
Indentation.

TFiFiE commented on src/Network.h:
```diff
@@ -104,18 +96,53 @@ class Network {
                                const std::vector<float>& V,
                                std::vector<float>& M, const int C, const int K);
+    Netresult get_scored_moves_internal(const GameState* const state,
+                                               const int symmetry);
```
Indentation.

TFiFiE commented on src/Network.h:
```diff
@@ -104,18 +96,53 @@ class Network {
-    static void forward_cpu(const std::vector<float>& input,
+    void forward_cpu(const std::vector<float>& input,
                             std::vector<float>& output_pol,
                             std::vector<float>& output_val);
```
Indentation and, again, you can make this member function **const**.

TFiFiE commented:
Would it perhaps be better for `Network::initialize` to be the constructor instead? Any instance of `Network` that can't be immediately initialized could be replaced by a `unique_ptr` to it.

ihavnoid commented:
@TFiFiE I agree with you but my opinion is that this PR is getting too huge, and hence let's go with another code restructuring PR to fix that one.  The problem (if we call it a 'problem') is actually everywhere so it better be addressed all at once.  There are constructor implications too which may add some complexity here and there.

ihavnoid commented:
@gcp  - any next steps?

gcp commented:
There were some small cleanups I wanted to try (use unique_ptr's instead of the raw ones for the global objects, and see if some of the globals can be made function scoped) but might as well merge this in already and put the cleanups on top of it.
